src/algorithms/tracking/SecondaryVertexFinder.cc

```diff
@@ -0,0 +1,414 @@
+
+      inputTracks.emplace_back(&(trajectory->trackParameters(tip)));
+      m_log->trace("Track local position at input = {} mm, {} mm",
+                   par.localPosition().x() / Acts::UnitConstants::mm,
+                   par.localPosition().y() / Acts::UnitConstants::mm);
+    }
+  }
+
+  std::vector<Acts::Vertex> vertices;
+  auto result = finder.find(inputTracks, vfOptions, stateSec);
+  if (result.ok()) {
+    vertices = std::move(result.value());
+  }
+
+  for (const auto& vtx : vertices) {
+    edm4eic::Cov4f cov(vtx.fullCovariance()(0, 0), vtx.fullCovariance()(1, 1),
+                       vtx.fullCovariance()(2, 2), vtx.fullCovariance()(3, 3),
+                       vtx.fullCovariance()(0, 1), vtx.fullCovariance()(0, 2),
+                       vtx.fullCovariance()(0, 3), vtx.fullCovariance()(1, 2),
+                       vtx.fullCovariance()(1, 3), vtx.fullCovariance()(2, 3));
+    auto eicvertex = prmVertices.create();
+    eicvertex.setType(1); // boolean flag if vertex is primary vertex of event
+    eicvertex.setChi2(static_cast<float>(vtx.fitQuality().first)); // chi2
+    eicvertex.setNdf(static_cast<float>(vtx.fitQuality().second)); // ndf
+    eicvertex.setPosition({
+        static_cast<float>(vtx.position().x()),
+        static_cast<float>(vtx.position().y()),
+        static_cast<float>(vtx.position().z()),
+        static_cast<float>(vtx.time()),
+    });                              // vtxposition
+    eicvertex.setPositionError(cov); // covariance
+
+    for (const auto& t : vtx.tracks()) {
+      const auto& par = vertexfinderConfigSec.extractParameters(t.originalParams);
+      m_log->trace("Track local position from vertex = {} mm, {} mm",
+                   par.localPosition().x() / Acts::UnitConstants::mm,
+                   par.localPosition().y() / Acts::UnitConstants::mm);
+      float loc_a = par.localPosition().x();
+      float loc_b = par.localPosition().y();
+
+      for (const auto& part : reconParticles) {
+        const auto& tracks = part.getTracks();
+        for (const auto& trk : tracks) {
+          const auto& traj    = trk.getTrajectory();
+          const auto& trkPars = traj.getTrackParameters();
+          for (const auto& par : trkPars) {
+            double EPSILON = std::numeric_limits<double>::epsilon(); // mm
+            if (std::abs((par.getLoc().a / edm4eic::unit::mm) - (loc_a / Acts::UnitConstants::mm)) <
+                    EPSILON &&
+                std::abs((par.getLoc().b / edm4eic::unit::mm) - (loc_b / Acts::UnitConstants::mm)) <
+                    EPSILON) {
+              m_log->trace(
+                  "From ReconParticles, track local position [Loc a, Loc b] = {} mm, {} mm",
+                  par.getLoc().a / edm4eic::unit::mm, par.getLoc().b / edm4eic::unit::mm);
+              eicvertex.addToAssociatedParticles(part);
+            } // endif
+          } // end for par
+        } // end for trk
+      } // end for part
+    } // end for t
```

Comment:
This whole matching is not ideal, and I'd like to see it removed instead of getting into our code base and left for others to refactor later.

The input to seed finding is (only) a set of track parameters. You have these track parameters as input in the form of ReconstructedParticles. But you ALSO pass and use the ActsExamples::Trajectories (which we are actively trying to get away from...). Then after seed finding you have to run this nested set of loops for matching with ReconstructedParticles. Why not start from ReconstructedParticles to fill your input tracks? Then when you get the vertices they will point to the tracks that are filled parallel to the ReconstructedParticles vector. There will be no matching loops necessary.

Reply:
We told @bishoyDHD that copy-pasting the https://github.com/eic/EICrecon/blob/246dae58831f605639f8086abab1a0b450d093cb/src/algorithms/tracking/IterativeVertexFinder.cc#L155-L182 is an acceptable workaround to the issue of associating to PODIO. I don't think writing a routine to convert ReconstructedParticles to Trajectories is a useful exercise, if we plan to move to PODIO for Acts EDM anyway. Even once we have that, there will be some matching loop to have.

Reply:
> I don't think writing a routine to convert ReconstructedParticles to Trajectories is a useful exercise.

Not what I'm suggesting. Look at the code. Instead of getTrackParameters from Trajectories, you can getTrackParameters from ReconstructedParticles, instead of doing exactly that in the loop later.

Reply:
Or the 3D position? But then we need to convert using bound surface.

Reply:
It's in the loop here in the segment that the comment is about:
```
   for (const auto& part : *reconParticles) { 
     const auto& tracks = part.getTracks(); 
     for (const auto& trk : tracks) { 
       const auto& traj    = trk.getTrajectory(); 
       const auto& trkPars = traj.getTrackParameters();
```
i.e. [edm4eic::ReconstructedParticle::getTracks](https://github.com/eic/EDM4eic/blob/main/edm4eic.yaml#L257), [edm4eic::Track::getTrajectory](https://github.com/eic/EDM4eic/blob/main/edm4eic.yaml#L468), [edm4eic::Trajectory::getTrackParameters](https://github.com/eic/EDM4eic/blob/main/edm4eic.yaml#L432)

Reply:
The idea is this: https://github.com/eic/EICrecon/pull/1992 (not even tested compilation, and has some remaining FIXMEs).

Reply:
> non-existent conversion routine

In the sense that CKFTracking does this already (minus surface lookup since it's a perigee that is re-created).